### ISSUE TITLE
Acl::SetRule performance is very slow

### ISSUE THREAD

**Issue by weierophinney** [laminas/laminas-permissions-acl#6]
We are dealing with rules in the ~~70k - 100k~~ 1k - 2k range.  It currently takes setRule around 2.5 seconds to process all of these rules.

I have been able to increase the performance of this method by roughly 40% by simply eliminating the use of array_merge.

Suggested Change Acl line:580 to:

```
foreach($children as $child) {
    $resources[] = $child;
}
```

Suggestions on how this method can be improved further would be great appreciated. The tree that is being generated takes approximately 100-130ms to serialize/unserialize.

Note: Acl configuration will be cached in Production.  

EDIT: 
Sorry rules are in the 1500 - 2000 range.  Roughly 300 roles and 300 resources.

---

Originally posted by @steverhoades at https://github.com/zendframework/zend-permissions-acl/issues/4

**Comment by weierophinney** [comment]
Here is the current benchmarks running 1538 rules.

```
Function                                        Count   Incl.       Excl.   Average (ms) 
Zend\Permissions\Acl\Acl::setRule()             1538    1,463.07    832.59  0.95    
Zend\Permissions\Acl\Acl::getRules()            70450   319.49      292.49  0.00    
Zend\Permissions\Acl\Acl::getResource()         70214   132.21      98.47   0.00    
Zend\Permissions\Acl\Acl::getChildResources()   70210   124.85      111.78  0.00    
Zend\Permissions\Acl\Acl::hasResource()         70808   34.23       34.17   0.00    

```


---

Originally posted by @steverhoades at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145318669

**Comment by weierophinney** [comment]
@steverhoades you can try patching the line you suggested, but you'd need to verify edge cases with string keys and array keys first.


---

Originally posted by @Ocramius at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145336365

**Comment by weierophinney** [comment]
@Ocramius good point.  
How about:

```
foreach($children as $key => $child) {
    $resources[$key] = $child;
}
```


---

Originally posted by @steverhoades at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145353100

**Comment by weierophinney** [comment]
@steverhoades can work, yes.


---

Originally posted by @Ocramius at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145355243

**Comment by weierophinney** [comment]
@steverhoades again, I'd start from the tests: the snippet above overwrites numeric keys.


---

Originally posted by @Ocramius at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145355383

**Comment by weierophinney** [comment]
@Ocramius I should know better, I'll check out the tests.  From reading the code it appears as  though the keys should be resourceIds... perhaps i have that wrong.


---

Originally posted by @steverhoades at https://github.com/zendframework/zend-permissions-acl/issues/4#issuecomment-145356275

**Comment by weierophinney** [comment]
@steverhoades I'm reviewing this now, and I think the area you reference in the original issue reads thus now:

```php
        $resourcesTemp = $resources;
        $resources = [];
        foreach ($resourcesTemp as $resource) {
            if (null !== $resource) {
                $resourceObj = $this->getResource($resource);
                $resourceId = $resourceObj->getResourceId();
                $children = $this->getChildResources($resourceObj);
                $resources = array_merge($resources, $children);
                $resources[$resourceId] = $resourceObj;
            } else {
                $resources[] = null;
            }
        }
        unset($resourcesTemp);
```

I'm trying to understand what your original suggestion was, exactly, as the above looks like necessary work for the most part. It kind of looked like you were suggesting the `$resources = array_merge($resources, $children)` line would become:

```php
foreach ($this->getChildResources($resourceObj) as $childId => $childResource) {
    $resources[$childId] = $childResource;
}
```

One thing to note: `getChildResources()` itself suffers from a similar issue, most likely.

I can likely make the change, but it would be good to benchmark. Is there any way we can get some sort of representative set from you to benchmark against?

**Comment by weierophinney** [comment]
I wrote the following benchmark, in the hopes it might demonstrate the issues:

```php
namespace Laminas\Permissions\Acl;

/**
 * @BeforeMethods({"init"})
 */
class LargeSetBench
{
    /** @var Acl */
    private $acl;

    /** @var string[] */
    private $resources;

    /** @var string[] */
    private $roles;

    public function init()
    {
        $this->acl   = $acl = new Acl();
        $this->roles = [];
        foreach (range('a', 'f') as $role) {
            $acl->addRole($role);
            foreach (range(0, 9) as $child) {
                $child = $role . (string) $child;
                $acl->addRole($child, $role);
            }
            $this->roles[] = $role;
        }

        $this->resources = [];
        foreach (range('A', 'F') as $resource) {
            $acl->addResource($resource);
            foreach (range(0, 9) as $child) {
                $child = $resource . (string) $child;
                $acl->addResource($child, $resource);
            }
            $this->resources[] = $resource;
        }
    }

    /**
     * @Revs(10)
     * @Iterations(5)
     */
    public function benchRules()
    {
        foreach (range(1, 2048) as $number) {
            $this->acl->setRule(Acl::OP_ADD, ACL::TYPE_ALLOW, $this->roles, $this->resources);
        }
    }
}
```

This essentially creates 66 roles (6 parents, each with 10 children), 66 resources (6 parents, each with 10 children), and 2048 rules (each assigned to the 6 parent roles and 6 parent resources).

I ran this against current code, and then modified code that refactors the `array_merge()` calls.

I observed a 3-5% increases in performance (I measured with PHP 7.2, as that would be the slowest). This was true both when using the benchmark as written, as well as when using a list of all roles and all resources (instead of only the parents) when setting rules.

As this is nowhere near your reported 40% improvement, I really need more information before I continue down this path, as I'm not sure I'm setting up a benchmark that accurately describes your scenario.

**Comment by steverhoades** [comment]
@weierophinney This issue is approximately 5-6 years old I think so was most likely run against PHP 5.3 or 5.4. Unfortunately I don't fully remember the exact use case nor do I have the code to reproduce as I'm no longer with that company.

The main issue was that we had very large ACL trees 100+ roles. I'll give this some further thought to see If I can recall anything further. Hope you're well!

**Comment by JacobBrownAustin** [comment]
Not sure if anyone is still following this issue since last comments are from over two years ago...
I made pull request to improve speeds for setRule() for our usage.  For our use case,  I've measured the method in our code base that uses all the ACL,  and it got 10x speed boost.
I just made a benchmark with phpbench to measure a similar use-case to ours (but with less calls to allow/deny because old cold was too slow),  and the benchmark is about 35x times faster with this pull request.
